hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+            LOG_DEBUG(TRACE, "BLE GAP event: security parameters request.");
+            // Pairing is not supported
+            ret = sd_ble_gap_sec_params_reply(event->evt.gap_evt.conn_handle, BLE_GAP_SEC_STATUS_PAIRING_NOT_SUPP, nullptr, nullptr);
+            if (ret != NRF_SUCCESS) {
+                LOG(ERROR, "sd_ble_gap_sec_params_reply() failed: %u", (unsigned)ret);
+            }
+        } break;
+        case BLE_GAP_EVT_AUTH_STATUS: {
+            LOG_DEBUG(TRACE, "BLE GAP event: authentication status updated.");
+            if (event->evt.gap_evt.params.auth_status.auth_status == BLE_GAP_SEC_STATUS_SUCCESS) {
+                LOG_DEBUG(TRACE, "Authentication succeeded");
+            } else {
+                LOG_DEBUG(WARN, "Authentication failed, status: %d", (int)event->evt.gap_evt.params.auth_status.auth_status);
+            }
+        } break;
+        case BLE_GAP_EVT_TIMEOUT: {
+            if (event->evt.gap_evt.params.timeout.src == BLE_GAP_TIMEOUT_SRC_AUTH_PAYLOAD) {
+                LOG_DEBUG(ERROR, "BLE GAP event: Authenticated payload timeout");
+            }
+        } break;
+        default: {
+            //LOG_DEBUG(TRACE, "Unhandled BLE event: %u", (unsigned)event->header.evt_id);
+        } break;
     }
-    chr->flags = halChar->flags;
-    halChar->handle = chr->handles.value_handle;
-    return 0;
 }
 
-int initService(ble_service* halService, Service* service) {
-    if (halService->char_count > BLE_MAX_CHAR_COUNT) {
-        LOG(ERROR, "Maximum number of characteristics exceeded");
-        return BLE_ERROR_INVALID_PARAM;
+/* FIXME: It causes a section type conflict if using NRF_SDH_BLE_OBSERVER() in c++ class.*/
+typedef struct {
+    BleObject::Broadcaster* instance;
+} BroadcasterImpl;
+static BroadcasterImpl broadcasterImpl;
+
+BleObject::Broadcaster::Broadcaster(BleObject* obj) : bleObj_(obj) {
+    isAdvertising_ = false;
+    advPending_ = false;
+    advHandle_ = BLE_GAP_ADV_SET_HANDLE_NOT_SET;
+    connHandle_ = BLE_INVALID_CONN_HANDLE;
+    /* Default advertising parameters. */
+    advParams_.version = 0x01;
+    advParams_.type = BLE_ADV_CONNECTABLE_SCANNABLE_UNDIRECRED_EVT;
+    advParams_.filter_policy = BLE_ADV_FP_ANY;
+    advParams_.interval = BLE_DEFAULT_ADVERTISING_INTERVAL;
+    advParams_.timeout = BLE_DEFAULT_ADVERTISING_TIMEOUT;
+    advParams_.inc_tx_power = false;
+    memset(advData_, 0x00, sizeof(advData_));
+    memset(scanRespData_, 0x00, sizeof(scanRespData_));
+    advDataLen_ = scanRespDataLen_ = 0;
+    /* Default advertising data. */
+    // FLags
+    advData_[advDataLen_++] = 0x02;
+    advData_[advDataLen_++] = BLE_SIG_AD_TYPE_FLAGS;
+    advData_[advDataLen_++] = BLE_SIG_ADV_FLAGS_LE_ONLY_GENERAL_DISC_MODE;
+    /* Default TX power. */
+    txPower_ = 0;
+}
+
+int BleObject::Broadcaster::init(void) {
+    int ret = configure(&advParams_);
+    CHECK_RETURN(ret, nullptr);
+    setTxPower(txPower_);
+    broadcasterImpl.instance = this;
```

Comment:
Why do we need this object? Can't we simply pass `this`? And even that is probably not necessary because `BleObject::instance()->xxx` can be used at all times.